### DISCUSSION TITLE
Cache miss although all the fields are cache hits

### DISCUSSION THREAD

**Discussion by villesau** [urql-graphql/urql#2867] [Q&A] [answered]
Hi,

I'm seeing urql cache miss for the query even though all the fields seem to be cache hits in the devtools. What could be the reason for cache miss even if all the individual fields are cache hits? I'm trying to get ssr working for the urql data, but it does not seem to help. See the screenshots for an example of what I'm looking at.

<img width="236" alt="Screenshot 2022-12-08 at 23 34 43" src="https://user-images.githubusercontent.com/2647016/206572369-99d3ea19-858d-4f08-9c32-417417450c78.png">
<img width="304" alt="Screenshot 2022-12-08 at 23 34 55" src="https://user-images.githubusercontent.com/2647016/206572376-6ab97375-da27-493e-9d7b-0032ff08f39d.png">

There's no field in that JSON that would indicate cache miss.

**Comment by JoviDeCroock** [comment, marked as answer]
It's a bit hard given these two screenshots to help you out here 😅

- what queries are you doing on the server/client?
- how are you restoring the cache?

Not entirely sure whether we forgot to mark the `ssrExchange` serving it as a miss, are you seeing a network request?

**Reply by villesau** [reply]
On the server I'm bundling queries to single query, but e.g the article query in question is returning the same fields + 1 more than on the client.

I'm restoring cache like this:

```
    ssrExchange({
      isClient: true,
      // @ts-expect-error __URQL_DATA__ not defined
      // eslint-disable-next-line no-underscore-dangle
      initialState: window.__URQL_DATA__
    }),
```
following instructions from here: https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/

and the __URQL_DATA__ is populated like:

```
<script>
  const data = // data populated here, trying even with hardcoded data
    window.__URQL_DATA__ = data ? JSON.parse(data) : undefined;
</script>
```

I am indeed seeing a network request, even with `cache-first` policy.

We do not have full blown SSR, so we are only rendering the data on server and trying to populate it on the client. So no React SSR in play here and that's why server & client are not 100% 1:1: On server isClient is set to false and more data is returned for html meta tags etc.

**Reply by JoviDeCroock** [reply]
> 1 more than on the client.

Well that's your reason, one more field will result in a different operation-key and the ssr-exchange is a key: response type of cache, transferring a normalized cache to the client would be pretty expensive, for ssr we assume identical flows i.e. same data.

**Reply by villesau** [reply]
Is the only option to do two queries for the article then? One for the identical data and one for the extra key? Like this:

```
  query GetArticle($articleId: ID!) {
    metaArticle: article(articleId: $articleId) {
      metaDescription
    }
    article(articleId: $articleId) {
        ...OnlyIdenticalFields
    }
 }
```

Or would that also affect caching?

**Reply by JoviDeCroock** [reply]
I mean, why would you do different queries on the client vs the server, my suggestion would be to just use the same queries 😅 btw your linked example is still one operation which means one operation-key which implies a fundamental difference from what you are describing in the above where you say you perform a different operation with one more field on the server

**Reply by villesau** [reply]
The reason why the queries are different is because I need less data on the (react) client: On the server when rendering the skeleton html we also render meta tags for search engines. On the client we only need to render the actual article content which we don't render on the server since our react setup does not stretch to SSR due to legacy reasons. I'm trying to find a middle ground for the first contentful paint: Not needing to go full SSR and see how only returning the relevant data affects that.

If the above is not sufficient, do I need to fire multiple individual queries?

**Reply by JoviDeCroock** [reply]
Ah I get it, well sorry to say this but you would need to do 1 query where you reflect the query you use in your UI and one where you leverage the `MetaDescription`.

One operation will result in the key you use later on in the UI and one will be unused

**Reply by villesau** [reply]
I now tried with exact 1:1 query. I still see a cache miss, but the request is not fired :+1: 

I'm wondering if I need to init two clients, one fully server side (could be even something simpler such as `graphql-request`) and one Urql for 1:1 the client data.

I still don't 100% understand why the request needs to be made if all the fields are found though. I guess the graph cache does not check the exact fields and whether those are found or not, but only the operations that's been made or something like that?

**Reply by JoviDeCroock** [reply]
It's the ssrExchange, it's a temp key-based cache and won't rehydrate ops it didn't see (first answer). Graphcache does 0 communication for rehydrate it just sees, oh I got a synchronous response from a later exchange.

In urql every operation+variables has a unique key, which is how the doc-cache works, we apply that principle for data rehydrate as well. Sending a normalized cache would be expensive as we can assume that applications render with the same queries ssr vs hydrate.

The miss in devtools is a bug, we aren't marking ssr results as cache hit which we should.

**Reply by villesau** [reply]
Ok, thanks for elaborating! I'm puzzled of the blazing fast response time 😅

**Reply by JoviDeCroock** [reply]
@villesau thank you for finding this bug for us, cache-ssr hits will be surfaced correctly now https://github.com/urql-graphql/urql/pull/2872
